### ISSUE TITLE
Implementation of ARIA slider role

### ISSUE THREAD

**Issue by chabbrik** [leongersen/noUiSlider#685]
It would be great if this slider would support accessibility. Right now no opensource slider implementation 
does it.

More details about slider role can be found there.
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_slider_role

**Comment by avprod1** [comment]
Hi, Leon. We could use this for united.com accessibility support. How can I help? I am looking at [https://refreshless.com/nouislider/examples/#section-keyboard](url)

**Comment by leongersen** [comment]
We'd have to label the dom elements with appropriate aria roles, and set the `aria-valuenow` on update. I'd need someone with experience in using a screen reader or similar software to verify that the roles are usable.

**Comment by leongersen** [comment]
That'd be nice.

**Comment by avprod1** [comment]
Hi, Leon.

I need to have this working in a few weeks.

I'd rather help you implement the change than start trying to dig through
jquery.nouislider.js and make local changes.

These articles are old, but may still be valid:
https://www.paciellogroup.com/blog/2008/05/aria-slider-part-1/
https://www.paciellogroup.com/blog/2008/06/aria-slider-part-2/

Suggests adding the role="slider" and other aria values.


How can I help the most?


--Scott
630-408-7588 cell
312298-4443 work

**Comment by chabbrik** [comment]
I have a working implementation of ARIA in nouislider, that we made for our customer. I am willing to integrate it into the core library, as it is mostly adding a few attributes to the nouislider tag.

However, I cannot find, where do I start. So if someone is willing to give me some pointers, I woud highly appreciate that.

I am attachning our workaround to the problem to illustrate how we did it.


[ARIA_code_suggestion.txt](https://github.com/leongersen/noUiSlider/files/600353/ARIA_code_suggestion.txt)

**Comment by avprod1** [comment]
Vasilij, where do I add this code block?


--Scott--

**Comment by chabbrik** [comment]
I got some time this week, so I will make an example implementation.

**Comment by chabbrik** [comment]
So the handle div should have following structure (based on 9.0.0 version):
`<div class="noUi-handle" data-handle="1" 
**tabindex="0" aria-valuemin="5" aria-valuenow="78.20" aria-valuemax="100" role="slider"** 
style="z-index: 4;">
<div class="noUi-tooltip">78.20</div>
</div>
`

The attributes in bold should be added to the handle div. I cannot find where in code the div is constructed unfortunately.

**Comment by chabbrik** [comment]
Updating aria-valuenow requires a JS code snippet somewhere:

```
slider.noUiSlider.on('end', function(values) {
    const handleLower = slider.querySelector('[data-handle="0"]');
    const handleUpper = slider.querySelector('[data-handle="1"]');
    handleLower.setAttribute('aria-valuenow', values[0]);
    handleUpper.setAttribute('aria-valuenow', values[1]);
});
```

If number of handles is different, then some sort of iterator is needed. Please give some pointers where in code it should be added and I will produce a pull request with working solutions.

**Comment by leongersen** [comment]
So, this identifies every handle as a separate slider?

**Comment by chabbrik** [comment]
It works with each handle as separate object, but not slider. Or I fail to understand the question.

**Comment by leongersen** [comment]
I've had a chance to look this up [in the spec](https://www.w3.org/TR/wai-aria-practices/#slidertwothumb), because I wasn't sure how to implement multiple handles. The spec says this, so @chabbrik has a correct implementation. I'll add this in the next revision of noUiSlider.

> Each element serving as a focusable slider thumb has role slider.

**Comment by leongersen** [comment]
- [ ] `tabindex`
- [ ] `aria-valuenow`: decimal value representing the current value of the slider;
- [ ] `aria-valuemin`: decimal value, minimum allowed;
- [ ] `aria-valuemax`: decimal value, maximum allowed;
- [ ] `aria-valuetext` a string that makes the slider value understandable, e.g., "Monday";
- [ ] `aria-labelledby`: needs to be added outside lib, add to docs;
- [ ] `aria-orientation`: `vertical`/`horizontal`;

**Comment by leongersen** [comment]
Initial aria support just shipped in noUiSlider 10.0.0. Feedback is very welcome!
